File: taojieTan/assignments/treesDFS/lc113/lc113.py
# Problem Statement #

# Given a binary tree and a number ‘S’, find all paths from root-to-leaf such that the sum of all the node values of each path equals ‘S’.

# Example
# -----
# targetSum = 12
# Output = [ [ 1, 7, 4 ], [ 1, 9, 2 ] ]
#      1
#     / \
#    7   9
#  / |   | \
# 4  5   2  7

'''
  edge case: empty tree => return []

  Overview:
  We can apply DFS pattern using stack (we use list) to find all root-to-leaf paths that are equal to target sum.
  Since leaf doesn't have left and right children, we can if condition to check if its left and right nodes are None, and the path sum is equal to target sum, 
  if yes, we store the path to final list; 
  else, we store its left or right node to stack, and keep repeating the steps until we traverse through every node. (when stack is empty).

  pseudo code:
  First declare and initialize a stack with (root node, target sum - root value, and sublist with root value), 
  and a empty result list to store paths that sum up tp target sum.
  while stack is not empty, 
  we pop the top value from stack, 
  and check if left and right nodes are None, and the path sum is equal to target sum,
  if yes, we append the sublist with path nodes into result list;
  else, we push the left node or right node, target sum - path sum, and sublist path node to stack.

  When finish the while loop, we will get the final result.
'''

class TreeNode:
  def __init__(self, val, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


def find_paths(root, target_sum):
  if root is None:
    return []

  allPaths = []
  node_stack = [(root, 0, [])]

  while len(node_stack):
    node, path_sum, path = node_stack.pop()

    if node.left is None and node.right is None and path_sum + node.val == target_sum:
      path.append(node.val)
      allPaths.append(path)
    
    if node.left:
      node_stack.append((node.left, path_sum + node.val, path + [node.val]))

    if node.right:
      node_stack.append((node.right, path_sum + node.val, path + [node.val]))

  return allPaths


def main():
  
  root1 = None
  sum = 23
  print("Tree paths with sum " + str(sum) + ": " + str(find_paths(root1, sum)))

  root = TreeNode(12)
  root.left = TreeNode(7)
  root.right = TreeNode(1)
  root.left.left = TreeNode(4)
  root.right.left = TreeNode(10)
  root.right.right = TreeNode(5)
  sum = 23
  print("Tree paths with sum " + str(sum) +
        ": " + str(find_paths(root, sum)))

'''
            12
          7    1
        4    10  5    => return [[12, 7, 4], [12, 1, 10]]
'''


main()

File: victorLi/assignments/treesDFS/lc113/lc113.js
// Problem Statement #

// Given a binary tree and a number ‘S’, find all paths from root-to-leaf such that the sum of all the node values of each path equals ‘S’.

// Example
// -----
// targetSum = 12
// Output = [ [ 1, 7, 4 ], [ 1, 9, 2 ] ]
//      1
//     / \
//    7   9
//  / |   | \
// 4  5   2  7

/*traverse using DFS
  if path = target then push stack into allPaths array



            12
          7    1
        4    10   5
                 5  3

          path 1 : 12 7 4
          path 2 : 12 1 10
          path 3 : 12 1 5 5     
*/

class TreeNode {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
};


const find_paths = function(root, sum) {
    allPaths = [];
    // TODO: Write your code here
    let prevSum = 0;
    let stack = []
    let sumStack = []
    let onePath = []
    
    let node = root;

    while( node || stack.length )
    {
      while( node )
      {
          stack.push(node)
          onePath.push(node.value)
          sumStack.push(prevSum += node.value)
          node = node.left
      }

      
      node = stack.pop()

    
    
      if( isLeaf(node) && prevSum === sum)
      {
          allPaths.push(onePath)
          console.log("-----ONEPATH--------------")
          console.log(onePath)
     
          
      }
      

      if( onePath.length > 1 &&  !node.right)
         onePath.pop()
      
  
      
      node = node.right;
      prevSum = sumStack.pop()

    }

    return allPaths;
};

let isLeaf = (node) =>{
  if( !node.right && !node.left )
    return true
  else
    return false
}

var root = new TreeNode(12)
root.left = new TreeNode(7)
root.right = new TreeNode(1)
root.left.left = new TreeNode(4)
root.right.left = new TreeNode(10)
root.right.right = new TreeNode(5)
root.right.right.left = new TreeNode(5)
root.right.right.right = new TreeNode(3)
sum = 23
console.log(`Tree paths with sum: ${sum}: ${find_paths(root, sum)}`)
